go-controller/pkg/factory/factory.go

```diff
@@ -468,18 +472,31 @@ func NewWatchFactory(c kubernetes.Interface, stopChan chan struct{}) (*WatchFact
 		}
 	}
 
+	wf.shutdownWg.Add(1)
+	startWg := sync.WaitGroup{}
+	startWg.Add(1)
 	go func() {
+		startWg.Done()
 		<-stopChan
 
 		// Remove all informer handlers
 		for _, inf := range wf.informers {
 			inf.shutdown()
 		}
+		wf.shutdownWg.Done()
 	}()
+	startWg.Wait()
```

Comment:
what is `startWg` for?

Reply:
@danwinship the goroutine to start

Reply:
yes, I mean, I can see that, but why does it matter?

Reply:
@danwinship caller (like a unit test?) calls NewWatchFactory(), gets returned factory, does something, calls shutdown. Goroutine finally starts and sits around blocking on the stopchannel, then when the stop channel finally closes, tries to call a bunch of stuff that's already torn down.

Reply:
But anyway, because of the removal of the stopChan we don't need the goroutine anymore.